nixos/modules/services/video/wivrn.nix

```
@@ -0,0 +1,137 @@
+        json = mkOption {
+          type = configFormat.type;
+          description = ''
+            Configuration for WiVRn. The attributes are serialized to JSON in config.json.
+            See https://github.com/Meumeu/WiVRn/blob/master/docs/configuration.md
+          '';
+          example = {
+            scale = 1.0;
+            bitrate = 100000000;
+            encoders = [
+              {
+                encoder = "nvenc";
+                codec = "h264";
+                width = 0.5;
+                height = 1.0;
+                offset_x = 0.0;
+                offset_y = 0.0;
+              }
+              {
+                encoder = "nvenc";
+                codec = "h264";
+                width = 0.5;
+                height = 1.0;
+                offset_x = 0.5;
+                offset_y = 0.0;
+              }
+            ];
+          };
+        };
+      };
+    };
+  };
+
+  imports = [
+    (mkAliasOptionModule [ "services" "wivrn" "monadoEnvironment" ] [ "systemd" "user" "services" "wivrn" "environment" ])
+  ];
+
+  config = mkIf cfg.enable {
+    security.wrappers."wivrn-server" = mkIf cfg.highPriority {
+      setuid = false;
+      owner = "root";
+      group = "root";
+      # cap_sys_nice needed for asynchronous reprojection
+      capabilities = "cap_sys_nice+eip";
+      source = getExe' cfg.package "wivrn-server";
+    };
+
+    systemd.user = {
+      services.wivrn = {
+        description = "WiVRn XR runtime service module";
+        requires = [ "wivrn.socket" ];
+        unitConfig.ConditionUser = "!root";
+        serviceConfig = {
+          ExecStart =
+            if cfg.highPriority
+            then "${config.security.wrapperDir}/wivrn-server" + optionalString cfg.config.enable " -f ${configFile}"
+            else getExe' cfg.package "wivrn-server" + optionalString cfg.config.enable " -f ${configFile}";
+          Restart = "no";
+        };
+        restartTriggers = [ cfg.package ];
+        wantedBy = [ "sockets.target" ];
+      };
+
+      sockets.wivrn = {
+        description = "WiVRn XR service module connection socket";
+        unitConfig.ConditionUser = "!root";
+        socketConfig = {
+          ListenStream = "%t/wivrn_comp_ipc";
+          RemoveOnStop = true;
+          # If WiVRn crashes while starting up, we want to close incoming OpenXR connections
+          FlushPending = true;
+        };
+        restartTriggers = [ cfg.package ];
+        wantedBy = [ "sockets.target" ];
+      };
+    };
+
+    services = {
+      wivrn.monadoEnvironment = {
+        # Default options
+        # https://gitlab.freedesktop.org/monado/monado/-/blob/4548e1738591d0904f8db4df8ede652ece889a76/src/xrt/targets/service/monado.in.service#L12
+        XRT_COMPOSITOR_LOG = mkDefault "debug";
+        XRT_PRINT_OPTIONS = mkDefault "on";
+        IPC_EXIT_ON_DISCONNECT = mkDefault "off";
+      };
+      udev.packages = with pkgs; [ xr-hardware ]; # WiVRn can be used with some wired headsets
```

```suggestion
      udev.packages = with pkgs; [ xr-hardware android-udev-rules ]; # WiVRn can be used with some wired headsets
```

Nit: it *probably* makes sense to install `android-udev-rules`. If you disagree, just mark this as resolved.

---

Yeah I didn't even think about that. Will add.

Like in the monado-vulkan-layers package in the other PR, would like you like me to remove the `with lib;` from meta?